### PR TITLE
Add test infrastructure for WebVR reftests

### LINKED ISSUE #22795: Support a test WebVR/XR display
For testing, it would be nice to have a WebVR/XR implementation that renders to the current window. Firefox has such a thing behind a `dom.vr.test.enabled` pref.

### LINKED ISSUE #22804: Add reftests for WebVR/XR
Once we have a test WebVR/XR display (#22795) we can write reftests which use it.

### PR DESCRIPTION
**Pull request by asajeffrey**
Adds a `dom.webvr.test` pref, which puts WebVR content into fullscreen mode, so we can use our shiny new fullscreen reftesting apparatus for writing WebVR reftests.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #22795 and fix #22804
- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22840)

### CODE REVIEW

**Comment by asajeffrey** [comment]
This PR is based on #22825 so should land after it does.

File: tests/wpt/mozilla/tests/mozilla/webvr/reftests/webvr-baseline.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE html>
+<html>
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/webvr/reftests/webvr-baseline.html]: Add the `reftest-wait` class to this, and remove it once rAF is done executing. This will wait until we've actually rendered things to take the reftest screenshot.

**Review by jdm — APPROVED** [review]

**Comment by asajeffrey** [comment]
As @jdm said on IRC (https://mozilla.logbot.info/servo/20190206#c15932019) we should get feedback from @paulrouget about whether we should enter fullscreen + use the browser's rAF for all devices without an external display, or whether we should add another property to the rust-webvr
API.

**Comment by asajeffrey** [comment]
Rebased. #22825 has landed, so this can land too, once @paulrouget has commented.

**Comment by asajeffrey** [comment]
Also cc @kearwood.

**Comment by asajeffrey** [comment]
IRC conversation with @paulrouget at https://mozilla.logbot.info/servo/20190207#c15934602

TL;DR: We should update rust-webvr to have three kinds of displays: external, internal-using-webrender, and internal-without-webrender. That last kind is used by oculusvr and googlevr.

**Comment by asajeffrey** [comment]
@paul: Added a flag to rust-webvr over in https://github.com/servo/rust-webvr/pull/49, and used it here to decide whether or not to do presentation inside servo.

File: components/script/dom/vrdisplay.rs [hunk 2]
```diff
@@ -561,13 +605,28 @@ impl VRDisplay {
...
+            None => return,
+        };
         let js_sender = self.global().script_chan();
         let address = Trusted::new(&*self);
         let near_init = self.depth_near.get();
         let far_init = self.depth_far.get();
         let pipeline_id = self.global().pipeline_id();
 
+        // For displays with no external display, we use the regular window
+        // rAF, so we don't need to spin up a dedicated thread.
```
Comment (paulrouget) [inline components/script/dom/vrdisplay.rs]: How hard would it be to make the TestDisplay sync_poses behave in a way that we would not need to have this specific code path? I'm asking because the test will cover less code by re-using the window's rAF.

File: components/script/dom/vrdisplay.rs [hunk 3]
```diff
@@ -496,6 +503,43 @@ impl VRDisplay {
...
+        }
+        if !self.display.borrow().capabilities.presented_by_browser {
+            self.layer_ctx.set(ctx);
+            return;
+        }
+
+        // Enter fullscreen mode.
+        // TODO: fullscreen mode isn't quite right because
+        // a) it requires the element to be attached to the DOM tree, and
+        // b) it fullscreens the element together with it's CSS styling.
```
Comment (paulrouget) [inline components/script/dom/vrdisplay.rs]: "its"

File: components/script/dom/vrdisplay.rs (hunk 2 above)
Comment (asajeffrey) [reply components/script/dom/vrdisplay.rs]: IRC conversation: https://mozilla.logbot.info/servo/20190208#c15940619

**Review by asajeffrey — COMMENTED** [review]
Addressed most of the comments, just got `sync_poses` to deal with.

**Comment by paulrouget** [comment]
Here are some questions:

- How will we handle a canvas that is not part of the DOM?
- What if another JS code triggers fullscreen mode?
- Beside the fact the code path doesn't exist yet, is there another reason to not bypass WebRender? (WebRender adds a little bit of overhead)
- Do you run Servo's compositor in the main thread? If so, can you move it in a separated thread and in immersive mode, run WebVR code in the main thread?

This general approach feels a bit like a work-around to me. And ML and FxR share a similar setup, so whatever we end-up doing for ML, we will do the same for FxR.

I think your approach might be incompatible with 2 parts of a longer term plan:
- use WebGL texture directly and pause Servo's compositor
- move RAF/VR calls to a WebVR process/thread, with shared texture (SurfaceTexture)

I'll be happy to r+ this as long as that we agree that it might be a temporary approach until we manage to make WebGL draw directly on a shared surface texture.

**Comment by MortimerGoro** [comment]
I talked with Paul this morning. Adding some feedback/questions:

- Using fullscreen approach may no be WebXR compliant. The API allows to use a canvas not added to the DOM, also offscreencanvas. What happens in those situations? We may hack it to add the canvas to the DOM but that would change the parentNode properties, or trigger other events which may not be spec compliant (Dom Listeners, fullscreenchanges). cc @kearwood  and @Manishearth may have an opinion about this)

- I'm concerned about the extra blit required by waiting for the Webrender Outout. IMO the goal should be to reduce the latency of the render path as much as possible. Right now we have 1 extra blit in the oculusvr/google vr platforms: `WebGL texture <-> SDK Eye Buffer FBO`. By using Webrender output I think we'd have 2 blits: `WebGL texture <-> WR Surface <-> SDK Eye Buffer FBO`. The ideal situation should be 0 extra blits. For example on Oculus, we can do direct render to a SurfaceTexture swapChain, making WebGL render directly to a Surface provided by the SDK.

`We should update rust-webvr to have three kinds of displays: external, internal-using-webrender, and internal-without-webrender. That last kind is used by oculusvr and googlevr.`

I Aaree IF we can't avoid the Magic Leap main thread limitations, other platforms should not be affected by this. For Firefox Reality backends I think the best render path would be to use the render target provided by the SDK (e.g. the SurfaceTexture swapChain from Oculus) and make WebGL render directly to that Surface, bypassing all the extra blits required on other situations.

**Comment by asajeffrey** [comment]
The goal of minimizing the overhead of each rAF loop seems like a good one to me. AFAICT there's two sources of overhead:

* Unecessary blit'ing. I'm confused as to why using WR requires an extra blit for devices like MagicLeap. WR is rendering into the surface provided by the embedder, so there isn't a separate `WR Surface <-> SDK Eye Buffer` blit. 

* WR. There is overhead from running WR, e.g. WR has to wait for layout to complete before rendering, even in cases like fullscreen or immersive mode where the DOM isn't contributing to the scene..

* Context switching between threads. (#22855)

I'm not sure what the trade-off is between these, e.g. how much of a bottleneck there is in waiting for WR.

**Comment by MortimerGoro** [comment]
`    * Unecessary blit'ing. I'm confused as to why using WR requires an extra blit for devices like MagicLeap. WR is rendering into the surface provided by the embedder, so there isn't a separate `WR Surface <-> SDK Eye Buffer` blit.`

The ideal path would be that WebGL renders to a surface provided by the VR SDK (e.g a swapChain on Oculus). That would be 0 blits.

Adding WR would add a extra blit in that situation., WebGL renders to a texture, and the the texture is blitted on the WR surface.

For  the 2 blits `WebGL texture <-> WR Surface <-> SDK Eye Buffer` I was thinking on platforms that don't allow creating swapChain to an Android Surface, and the "WebView" lives in another process from the Main app that renders the VR UI (e.g. GeckoView lives in another process in Firefox Reality)

**Comment by paulrouget** [comment]
Should we keep this PR open?

**Comment by asajeffrey** [comment]
This PR can close, but there may still be an issue around WebXR ref tests.@Manishearth, are there any WebXR ref tests, or are they all tests of the JS APIs?

**Comment by Manishearth** [comment]
JS APIs only. A Webxr reftest would be interesting.
